ffi_how_to.md

```diff
@@ -0,0 +1,115 @@
+    // When building for darwin/arm64, Go tooling automatically selects this file due to the specified build 
+    // tags. It then fetches the "prebuilt-ffi-darwin-arm64" module from the module proxy, using the version 
+    // specified in the `go.mod` file.
+)
+
+// Hash computes the digest of a message
+func Hash(message Message) Digest {
+	return prebuilt.Hash(message)
+}
+```
+
+```go
+prebuilt_bls_linux_amd64.go
+//go:build cgo && linux && amd64 && !ffi_source
+// +build cgo,linux,amd64
+// +build !ffi_source
+package ffi
+import (
+	prebuilt "fil.org/prebuilt-ffi-linux-amd64"
+)
+
+// Hash computes the digest of a message
+func Hash(message Message) Digest {
+	return prebuilt.Hash(message)
+}
+```
+
+```go
+// Build from source
+bls__source.go
+//go:build ffi_source
+// +build ffi_source
+// Same code as we have today
+```
+
+And so on and so forth for each (GOOS + GOARCH) combination for each of those files containing public APIs that currently call into the CGO bindings.
+
+The logical next question to ask is how are the `prebuilt-ffi-{GOOS}-{GOARCH}` modules mentioned here created and where are they hosted?
+
+### Building and publishing the prebuit-ffi modules + CI changes
+
+For each release of `filecoin-ffi`, CI will build and publish the corresponding `prebuilt-ffi-{GOOS}-{GOARCH}` modules as "go mod compatible zip files" to the Github release assets page for `filecoin-ffi` for each supported combination of (GOOS + GOARCH).
+
+In addition to the pre-built zip modules, we will also need to publish the corresponding `go.mod` and "info" meta for each pre-built module (go tooling needs these -> more details in the `Go Module Proxy` section below) . These can be created synthetically and can be persisted in the Github release assets as well.
+
+
+We already have some flavour of this today. See the `Assets` section [example](https://github.com/filecoin-project/filecoin-ffi/releases/tag/ed08caaf8778e1b6).
+
+
+The high level steps to create these assets for each `prebuilt-ffi-{GOOS}-{GOARCH}` module are as follows.
+For each release `vx.y.z` for `filecoin-ffi`:
+
+1. Clone `filecoin-ffi-vx.y.z` source on a machine with {GOOS}X and {GOARCH}Y.
+2. Remove all the `prebuilt_*` files
+3. Build it from source to create the prebuilt assets
+4. Remove all the transient build assets in `rust/target` dir
+5. Create the `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.info` and `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.mod` files (the latter can be created by removing the existing `go.mod` file and running `go mod tidy` to generate a new one)
+5. Zip it up using something like `https://github.com/aarshkshah1992/prebuilt-ffi-zipper` (the directories inside the zip just need to follow a specific hierarchy)
+5. Publish the `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.zip`, `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.mod` and `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.info` files to the Github release assets page for the `filecoin-ffi` repo
```

Comment:
Another dumb question: does:
1. does `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.zip` contain `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.mod` and `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.info` AND
2. do both `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.mod` and `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.info` get published alongside `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.zip`
?

Taking `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.mod` as example, this means there is one copy in `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.zip` and one copy uploaded alongside `prebuilt_bls_{GOOS}_{GOARCH}-{vx.y.z}.zip`.